### PR TITLE
GH-331: As a smith I need a Data-Flow analyses engine (with warnings if a ref is used before initialization) (Part 2)

### PR DESCRIPTION
**Pull request by mmews-n4**
This is the second PR of #331, but not the final one (successor of #390).

### CODE REVIEW

File: plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/factories/CFEEffectInfos.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+import org.eclipse.n4js.n4JS.IndexedAccessExpression;
+import org.eclipse.n4js.n4JS.ParameterizedCallExpression;
+import org.eclipse.n4js.n4JS.ParameterizedPropertyAccessExpression;
+import org.eclipse.n4js.n4JS.PostfixExpression;
+import org.eclipse.n4js.n4JS.UnaryExpression;
+import org.eclipse.n4js.n4JS.UnaryOperator;
+import org.eclipse.n4js.n4JS.VariableDeclaration;
+import org.eclipse.n4js.n4JS.util.N4JSSwitch;
+
+class CFEEffectInfos {
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/factories/CFEEffectInfos.java]: is it required to prefix non public class with all the acronyms?

File: plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/model/ControlFlowEdge.java [hunk 2]
```diff
@@ -36,7 +36,7 @@ public ControlFlowEdge(Node start, Node end) {
 	 */
 	public ControlFlowEdge(Node start, Node end, ControlFlowType cfType) {
 		super(start, end);
-		assert cfType.isBackwards() || start.astPosition < end.astPosition : "Edge has wrong direction";
+		assert cfType.isBackwards() == start.astPosition > end.astPosition : "Edge has wrong direction";
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/model/ControlFlowEdge.java]: a bit of nesting here and operator precedence is happening here. Could you add some `(` and `)` ? Like one pair at least?

Comment (mmews-n4) [reply plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/model/ControlFlowEdge.java]: done

File: plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/analysers/DeadCodeAnalyser.java [hunk 3]
```diff
@@ -146,17 +147,18 @@ public boolean isDeadCode(ControlFlowElement cfe) {
 	}
 
 	/** Finds the nearest reachable {@link Block} of the given {@link ControlFlowElement} */
-	private Block getReachableBlock(Set<ControlFlowElement> unreachableElems, ControlFlowElement unreachableElem,
+	private EObject getReachableBlock(Set<ControlFlowElement> unreachableElems, ControlFlowElement unreachableElem,
```
[the remark this answers is not in the document]
Comment (mmews-n4) [reply plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/analysers/DeadCodeAnalyser.java]: done

**Comment by mmews-n4** [comment]
Thanks!

This PR does not close #331.

**Pull request merged**